codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -0,0 +1,438 @@
+import software.amazon.smithy.model.shapes.ShapeType;
+import software.amazon.smithy.model.traits.EnumTrait;
+import software.amazon.smithy.model.traits.HttpTrait;
+import software.amazon.smithy.model.traits.TimestampFormatTrait.Format;
+import software.amazon.smithy.utils.OptionalUtils;
+
+
+/**
+ * Abstract implementation useful for all protocols that use HTTP bindings.
+ */
+public abstract class HttpBindingProtocolGenerator implements ProtocolGenerator {
+
+    private static final Logger LOGGER = Logger.getLogger(HttpBindingProtocolGenerator.class.getName());
+
+    private final boolean isErrorCodeInBody;
+
+    /**
+     * Creates a Http binding protocol generator.
+     *
+     * @param isErrorCodeInBody A boolean that indicates if the error code for the implementing protocol is located in
+     *                          the error response body, meaning this generator will parse the body before attempting to
+     *                          load an error code.
+     */
+    public HttpBindingProtocolGenerator(boolean isErrorCodeInBody) {
+        this.isErrorCodeInBody = isErrorCodeInBody;
+    }
+
+    @Override
+    public ApplicationProtocol getApplicationProtocol() {
+        return ApplicationProtocol.createDefaultHttpApplicationProtocol();
+    }
+
+    @Override
+    public void generateSharedComponents(GenerationContext context) {
+        // pass
+    }
+
+    @Override
+    public void generateRequestSerializers(GenerationContext context) {
+        TopDownIndex topDownIndex = context.getModel().getKnowledge(TopDownIndex.class);
+
+        Set<OperationShape> containedOperations = new TreeSet<>(
+                topDownIndex.getContainedOperations(context.getService()));
+        for (OperationShape operation : containedOperations) {
+            OptionalUtils.ifPresentOrElse(
+                    operation.getTrait(HttpTrait.class),
+                    httpTrait -> generateOperationSerializer(context, operation),
+                    () -> LOGGER.warning(String.format(
+                            "Unable to generate %s protocol request bindings for %s because it does not have an "
+                                    + "http binding trait", getProtocol(), operation.getId())));
+        }
+    }
+
+    @Override
+    public void generateResponseDeserializers(GenerationContext context) {
+        // pass
+    }
+
+    /**
+     * Gets the default serde format for timestamps.
+     *
+     * @return Returns the default format.
+     */
+    protected abstract Format getDocumentTimestampFormat();
+
+    /**
+     * Gets the default content-type when a document is synthesized in the body.
+     *
+     * @return Returns the default content-type.
+     */
+    protected abstract String getDocumentContentType();
+
+    private void generateOperationSerializer(
+            GenerationContext context,
+            OperationShape operation
+    ) {
+        generateOperationHttpBindingSerializer(context, operation);
```

Comment:
Why the redirection here? Both functions have the same signature

Reply:
Currently in later PRs I will be publishing this method also calls various other helper functions for constructing middleware and delegating the generation of document serializer shapes.